src/framing/FrameTransport.cpp

```diff
@@ -175,62 +168,60 @@ void FrameTransport::request(int64_t n) noexcept {
     // stack.
     return;
   }
-  drainOutputFramesQueue();
+
+  drainWrites(lock);
 }
 
-void FrameTransport::cancel() noexcept {
+void FrameTransport::cancel() {
   VLOG(6) << "cancel";
-  terminateFrameProcessor(folly::exception_wrapper());
+  terminateProcessor(folly::exception_wrapper());
 }
 
 void FrameTransport::outputFrameOrEnqueue(std::unique_ptr<folly::IOBuf> frame) {
-  std::lock_guard<std::recursive_mutex> lock(mutex_);
+  Lock lock(mutex_);
 
-  // We allow sending frames even without a frame processor so it's possible
-  // to send terminal frames without expecting anything in return
+  // We allow sending frames even without a frame processor so it's possible to
+  // send terminal frames without expecting anything in return.
   if (connection_) {
-    drainOutputFramesQueue();
+    drainWrites(lock);
     if (pendingWrites_.empty() && writeAllowance_.tryAcquire()) {
-      // TODO: temporary disabling VLOG as we don't know the correct
-      // frame serializer here. There is refactoring of this class planned
-      // which will allow enabling it again.
-      // VLOG(3) << this << " writing frame " << FrameHeader::peekType(*frame);
-
       connectionOutput_->onNext(std::move(frame));
       return;
     }
   }
-  // TODO: temporary disabling VLOG as we don't know the correct
-  // frame serializer here. There is refactoring of this class planned
-  // which will allow enabling it again.
-  // VLOG(3) << this << " queuing frame " << FrameHeader::peekType(*frame);
 
-  // We either have no allowance to perform the operation, or the queue has
-  // not been drained (e.g. we're looping in ::request).
-  // or we are disconnected
+  // We either have no allowance to perform the operation, or the queue has not
+  // been drained (e.g. we're looping in ::request), or we are disconnected.
   pendingWrites_.emplace_back(std::move(frame));
 }
 
-void FrameTransport::drainOutputFramesQueue() {
-  std::lock_guard<std::recursive_mutex> lock(mutex_);
+void FrameTransport::drainReads(const FrameTransport::Lock&) {
```

Comment:
I liked it that we pass a Lock inside and force this method to be single threaded without locking the recursive_mutex again. 

But I wonder, if we give different locked mutexes to this function, then this function will not be single threaded. So it is not the same as the original code. Namely, now we use a single mutex object and it is not a problem, but might passing this parameter in or writing a comment have the same result?

Do you think with your current update, is it possible to change recursive_mutex to non-recursive mutex?